### PR TITLE
Processor refactoring

### PR DESCRIPTION
**Pull request by thibault-martinez**
- Removes a useless dynamic allocation that caused a leak
- Adds transaction validation
- Changes the queue to a utash-based queue + cond/lock
- Removes dependency to core
- Some performance improvements
- Renames/cleans/documents

# Test Plan:
CI must pass.

### CODE REVIEW

File: gossip/components/processor.c [hunk 1]
```diff
@@ -7,269 +7,384 @@
...
+#define PROCESSOR_LOGGER_ID "processor"
+#define PROCESSOR_TIMEOUT_SEC 5
 
-static retcode_t process_transaction_bytes(processor_state_t *const state,
+/*
+ * Private functions
+ */
+
+/**
+ * Converts transaction bytes from a packet to a transaction, validates it and
```
Comment (th0br0) [inline gossip/components/processor.c:23]: Thanks for adding the documentation! :)

File: gossip/components/processor.c [hunk 2]
```diff
@@ -7,269 +7,384 @@
...
-
-    flex_trits_from_bytes(tx_flex_trits, NUM_TRITS_SERIALIZED_TRANSACTION,
-                          packet->content, NUM_TRITS_SERIALIZED_TRANSACTION,
-                          NUM_TRITS_SERIALIZED_TRANSACTION);
-    if ((*tx = transaction_deserialize(tx_flex_trits)) == NULL) {
-      neighbor->nbr_invalid_tx++;
-      log_warning(PROCESSOR_COMPONENT_LOGGER_ID,
-                  "Deserializing transaction failed\n");
-      return RC_PROCESSOR_COMPONENT_INVALID_TX;
+  // TODO Check if transaction hash is cached
```
Comment (th0br0) [inline gossip/components/processor.c:51]: rel #471 (just linking)

File: gossip/components/processor.c [hunk 3]
```diff
@@ -7,269 +7,384 @@
...
+
+  if (false) {
```
Comment (th0br0) [inline gossip/components/processor.c]: remove

File: gossip/components/processor.c [hunk 4]
```diff
@@ -7,269 +7,384 @@
...
+      log_warning(PROCESSOR_LOGGER_ID, "Invalid transaction bytes\n");
+      ret = RC_PROCESSOR_INVALID_TRANSACTION;
+      goto failure;
     }
-    // TODO(thibault): Transaction validation
-    // TODO(thibault): Add to cache
 
-    TRIT_ARRAY_DECLARE(hash, NUM_TRITS_HASH);
-    trit_array_set_trits(&hash, (*tx)->hash, NUM_TRITS_HASH);
+    // Deserializes th transaction
```
Comment (th0br0) [inline gossip/components/processor.c]: `the`

File: gossip/components/processor.c [hunk 5]
```diff
@@ -7,269 +7,384 @@
...
 
+    // Validates the transaction
+    if (!iota_consensus_transaction_validate(processor->transaction_validator,
+                                             &transaction)) {
+      log_warning(PROCESSOR_LOGGER_ID, "Invalid transaction\n");
+      ret = RC_PROCESSOR_INVALID_TRANSACTION;
+      goto failure;
+    }
+
+    // TODO Add transaction hash to cache
```
Comment (th0br0) [inline gossip/components/processor.c]: This is a recently-seen-bytes cache, independent of #471

rel #153

File: gossip/components/processor.c [hunk 6]
```diff
@@ -7,269 +7,384 @@
...
-      // Broadcast new transaction
-      log_debug(PROCESSOR_COMPONENT_LOGGER_ID,
-                "Propagating packet to broadcaster\n");
+
+      // Checks solidity of the transaction
       if ((ret =
-               broadcaster_on_next(&state->node->broadcaster, tx_flex_trits))) {
-        log_warning(PROCESSOR_COMPONENT_LOGGER_ID,
-                    "Propagating packet to broadcaster failed\n");
+               iota_consensus_transaction_solidifier_check_and_update_solid_state(
```
Comment (th0br0) [inline gossip/components/processor.c]: Will this also cause a traversal? If so, this is a potential DoS vector

File: gossip/components/processor.c [hunk 7]
```diff
@@ -7,269 +7,384 @@
...
-                  "Deserializing transaction failed\n");
-      return RC_PROCESSOR_COMPONENT_INVALID_TX;
+  // TODO Check if transaction hash is cached
+
+  if (false) {
+  } else {
+    bool exists = false;
+    struct _iota_transaction transaction = {.snapshot_index = 0, .solid = 0};
+    flex_trit_t transaction_flex_trits[FLEX_TRIT_SIZE_8019];
+    trit_array_t key = {.trits = hash,
```
Comment (th0br0) [inline gossip/components/processor.c]: As far as I can tell, you're never computing the hash and thus never writing any values to `hash`

File: gossip/iota_packet.h [hunk 8]
```diff
@@ -8,56 +8,135 @@
...
 
-// Forward declarations
-typedef struct _iota_transaction* iota_transaction_t;
-typedef struct _trit_array* trit_array_p;
-
+/**
+ * The IOTA gossip protocol exchange packet that contains:
+ * - A transaction encoded in bytes
+ * - A request hash encoded in bytes
+ * - An endpoint containing information related to the packet sender
```
Comment (th0br0) [inline gossip/iota_packet.h]: Maybe say that this is metadata ?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: gossip/components/processor.c (hunk 7 above)
Comment (th0br0) [reply gossip/components/processor.c]: Nevermind, below is a `memcpy` :sweat_smile:

**Review by th0br0 — APPROVED** [review]

File: gossip/components/processor.c (hunk 6 above)
Comment (thibault-martinez) [reply gossip/components/processor.c]: No, just checking solidity of current tx, branch and trunk

**Pull request merged**